# plot_stranded_coverage_multinorm.py
Python script used to calculate and plot normalized strand-specific coverage across discrete regions of the genome (e.g., individual open reading frames) 

### Dependencies
- python
- pysam (https://github.com/pysam-developers/pysam)
- samtools (http://www.htslib.org/)
- numpy (www.numpy.org)
- scipy (https://www.scipy.org/)
- matplotlib (https://matplotlib.org)


### Input

1. One or more **sam or bam files**
    -Can either be: 
        a. specified as arguments on the command line 
        b. provided in a text file (-b) as absolute or relative paths

2. A **tab-delimited annotations file** with one row per region, 4 columns and **NO header**:
    1. Reference name
    2. Region name (arbitrary string)
    3. First nucleotide position included in the region (0-based)
    3. Last nucleotide position included in the region (0-based)

### Usage

To get get usage info:
```
plot_stranded_coverage_multinorm_v#.#.py  -h
```

Specifying bams in a text file (-b) and utilizing only R1 (default) with mapping quality ≥ 30 (-q 30). Confidence intervals are standard deviation (default):
```
plot_stranded_coverage_multinorm_v#.#.py -b bams_info.txt -a regions.txt -o out -q 30
```

Specifying bams as arguments on the command line and utilizing only R2 (--useR2) with mapping quality ≥ 30 (-q 30). Confidence intervals are standard error (--sem):
```
plot_stranded_coverage_multinorm_v#.#.py -a regions.txt -o out -q 30 --useR2 --sem sample1.bam sample2.bam sample3.bam
```


### Output

    Note: 
    - If using R1, Forward reads are from RNA molecules that were the reverse complement of the reference (E.g., if using positive strand reference, R1 Forward reads = negative strand)
    - If using R2, Forward reads are from RNA molecules that were the same strand as the reference (E.g., if using positive strand reference, R2 Forward reads = positive strand)

    1. out_ref.pdf
        -Normalized coverage plot generated by matplotlib
            - Red = Forward strand coveage
            - Green = Reverse strand coverage
        - ref = name of reference sequence used for alignment

    2. Tab-delimited file with the normalized values used to generate the plot. Includes 1 header row and 1 row for each discrete region specified in the annotations file
        - Each row will contain 6 columns:
            1. Name of region (as specified in annotations file)
            2. x-coordinate used for plotting (middle of discrete region)
            3. Average normalized reverse strand coverage
            4. Standard deviation (or error, as specified) of the average normalized reverse strand coverage
            5. Average normalized forward strand coverage
            6. Standard deviation (or error, as specified) of the average normalized forward strand coverage
        
### Options

  ```
Options:
  -h, --help            show this help message and exit
  -b BAMS, --bams=BAMS  File with a bunch of bam file locations. Can be used
                        instead of providing bams as arguments [None]
  -a ANNOTS, --annots=ANNOTS
                        File with names and coordinates for regions to include
                        in plot (0-based) [None]
  -o OUT, --out=OUT     Base name for output files, both a pdf plot and the
                        values used in the plot [None]
  -q MAPQ, --mapq=MAPQ  minimum mapping quality to be used [20]
  --sem                 Display confidence intervals as standard error, as
                        opposed to standard deviation [false]
  --useR2               Use this flag if you want use R2 instead of R1 [false]
```



Copyright (C) 2017  Jason Ladner

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
