# chimeric_reads.py
Python script used to identify chimeric reads starting from a sam/bam alignmnet file generated by BWA mem.


### Dependencies
- python
- pysam (https://github.com/pysam-developers/pysam)
- samtools (http://www.htslib.org/)
- numpy (www.numpy.org)
- scipy (https://www.scipy.org/)
- matplotlib (https://matplotlib.org)

### Input

- One or more **sam/bam files** generated using BWA mem (https://github.com/lh3/bwa)
    - If only analyzing a single aligment, the sam/bam can be specified with '-b' and the base string to use for output files should be specified with '-o'
    - If batch analyzing multiple alignemts, specify alignments with '-i'. Please provide a tab-delimited text file with 2 columns and NO header:
        1. sam/bam file name 
        2. base string for output files.
    
### Usage

```
Note: "~/GDrive/scripts/" should be replaced with the path to the script on your system. And "v#.#" should be replaced with the version number being used (e.g., v3.6.2)
```

To get get usage info:
```
~/GDrive/scripts/chimeric_reads_v#.#.py  -h
```

To analyze a single bam file:
```
~/GDrive/scripts/chimeric_reads_v#.#.py  -b file1.bam  -o outname1
```

To analyze multiple bam files:
```
~/GDrive/scripts/chimeric_reads_v#.#.py  -i tabdelim_baminfo.txt
```

To analyze multiple bam files with strand-specific data:
```
~/GDrive/scripts/chimeric_reads_v#.#.py  -i tabdelim_baminfo.txt  --dataStranded
```

To analyze multiple bam files with strand-specific data and to filter for putative duplicate reads based on mapping coordinates:
```
~/GDrive/scripts/chimeric_reads_v#.#.py  -i tabdelim_baminfo.txt  --dataStranded --filterDups
```


### Output

    Coming soon!
    
### Options

  ```
  -h, --help            show this help message and exit
  -i INPUT, --input=INPUT
                        File with info for bulk processing. Should be tab-
                        delimited, 2 columns: 1) sam/bam file and 2) base
                        string for output files. Need to provide either -b and
                        -o or -i [None, OPT]
  -b BAM, --bam=BAM     input sam or bam file. Need to provide either -b
                        and -o or -i [None, OPT]
  -o OUT, --out=OUT     Base string for output files. Will only be used with
                        -b option for running a single sam/bam. Need to
                        provide either -b and -o or -i [None, OPT]
  -m MINPERC, --minPerc=MINPERC
                        Minumum percent of query read that must be aligned to
                        the reference, when considering all alignemnts. [0.99]
  -f BUFF, --buff=BUFF  Used to expand the x and y limits of the plots beyond
                        the beginning and end of the references. The buffer is
                        applied equally to min/max and x/y. [0]
  --filterDups          Use this flag if you do not want to include putative
                        duplicate counts in the figures that are generated.
                        [False]
  --dataStranded        Use this flag if your data is strand-specific. [False]  ```



Copyright (C) 2017  Jason Ladner

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
